File: backend/src/utils/structureMatching.ts
/**
 * Structure Matching Utility
 *
 * Computes structure preservation scores between original sigil control image
 * and AI-generated output. Uses pixel overlap and edge comparison.
 *
 * This replaces the placeholder Math.random() scores with real metrics.
 */

import sharp from 'sharp';
import https from 'https';
import http from 'http';
import { logger } from './logger';

/**
 * Structure match result
 */
export interface StructureMatchResult {
  iouScore: number;           // Intersection over Union (0-1)
  edgeOverlapScore: number;   // Edge-based overlap (0-1)
  combinedScore: number;      // Weighted combination (0-1)
  structurePreserved: boolean;
  classification: 'Structure Preserved' | 'More Artistic' | 'Style Drift';
  analysisDetails: {
    originalWhitePixels: number;
    generatedWhitePixels: number;
    intersectionPixels: number;
    unionPixels: number;
  };
}

/**
 * Thresholds for classification
 */
const THRESHOLDS = {
  preserved: 0.85,
  artistic: 0.70,
};

/**
 * Download image from URL and return as Buffer
 */
async function downloadImage(url: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith('https') ? https : http;

    protocol.get(url, (response) => {
      // Handle redirects
      if (response.statusCode === 301 || response.statusCode === 302) {
        const redirectUrl = response.headers.location;
        if (redirectUrl) {
          downloadImage(redirectUrl).then(resolve).catch(reject);
          return;
        }
      }

      if (response.statusCode !== 200) {
        reject(new Error(`Failed to download image: ${response.statusCode}`));
        return;
      }

      const chunks: Buffer[] = [];
      response.on('data', (chunk) => chunks.push(chunk));
      response.on('end', () => resolve(Buffer.concat(chunks)));
      response.on('error', reject);
    }).on('error', reject);
  });
}

/**
 * Convert image to grayscale and extract binary mask
 * White pixels (>threshold) become 1, others become 0
 */
async function extractBinaryMask(
  imageBuffer: Buffer,
  threshold: number = 128,
  targetSize: number = 512
): Promise<{ mask: Uint8Array; width: number; height: number }> {
  // Resize and convert to grayscale
  const { data, info } = await sharp(imageBuffer)
    .resize(targetSize, targetSize, { fit: 'contain', background: { r: 0, g: 0, b: 0 } })
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  // Create binary mask
  const mask = new Uint8Array(data.length);
  for (let i = 0; i < data.length; i++) {
    mask[i] = data[i] > threshold ? 1 : 0;
  }

  return { mask, width: info.width, height: info.height };
}

/**
 * Extract sigil structure from generated image using adaptive thresholding
 * Generated images have textures/colors, so we need smarter extraction
 */
async function extractGeneratedSigilMask(
  imageBuffer: Buffer,
  targetSize: number = 512
): Promise<{ mask: Uint8Array; width: number; height: number }> {
  // Convert to grayscale
  const { data, info } = await sharp(imageBuffer)
    .resize(targetSize, targetSize, { fit: 'contain', background: { r: 0, g: 0, b: 0 } })
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  // Calculate histogram to find optimal threshold (Otsu's method simplified)
  const histogram = new Array(256).fill(0);
  for (let i = 0; i < data.length; i++) {
    histogram[data[i]]++;
  }

  // Find threshold using mean of non-black pixels
  let sum = 0;
  let count = 0;
  for (let i = 10; i < 256; i++) {  // Skip very dark pixels
    sum += i * histogram[i];
    count += histogram[i];
  }
  const meanBrightness = count > 0 ? sum / count : 128;

  // Use threshold slightly below mean
  const threshold = Math.max(30, meanBrightness * 0.6);

  // Create binary mask
  const mask = new Uint8Array(data.length);
  for (let i = 0; i < data.length; i++) {
    mask[i] = data[i] > threshold ? 1 : 0;
  }

  return { mask, width: info.width, height: info.height };
}

/**
 * Compute Intersection over Union (IoU) between two binary masks
 */
function computeIoU(mask1: Uint8Array, mask2: Uint8Array): {
  iou: number;
  intersection: number;
  union: number;
  mask1Pixels: number;
  mask2Pixels: number;
} {
  let intersection = 0;
  let union = 0;
  let mask1Pixels = 0;
  let mask2Pixels = 0;

  for (let i = 0; i < mask1.length; i++) {
    const a = mask1[i];
    const b = mask2[i];

    mask1Pixels += a;
    mask2Pixels += b;

    if (a === 1 && b === 1) {
      intersection++;
    }
    if (a === 1 || b === 1) {
      union++;
    }
  }

  const iou = union > 0 ? intersection / union : 0;

  return { iou, intersection, union, mask1Pixels, mask2Pixels };
}

/**
 * Compute edge-based overlap with tolerance
 * Dilates the original mask slightly before comparison
 */
function computeEdgeOverlap(
  originalMask: Uint8Array,
  generatedMask: Uint8Array,
  width: number,
  height: number,
  tolerancePx: number = 3
): number {
  // Dilate original mask to allow for slight position differences
  const dilatedMask = new Uint8Array(originalMask.length);

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = y * width + x;

      // Check if any pixel in the tolerance radius is set
      let found = false;
      for (let dy = -tolerancePx; dy <= tolerancePx && !found; dy++) {
        for (let dx = -tolerancePx; dx <= tolerancePx && !found; dx++) {
          const ny = y + dy;
          const nx = x + dx;
          if (ny >= 0 && ny < height && nx >= 0 && nx < width) {
            const nidx = ny * width + nx;
            if (originalMask[nidx] === 1) {
              found = true;
            }
          }
        }
      }
      dilatedMask[idx] = found ? 1 : 0;
    }
  }

  // Count how many generated pixels fall within dilated original
  let covered = 0;
  let totalGenerated = 0;

  for (let i = 0; i < generatedMask.length; i++) {
    if (generatedMask[i] === 1) {
      totalGenerated++;
      if (dilatedMask[i] === 1) {
        covered++;
      }
    }
  }

  return totalGenerated > 0 ? covered / totalGenerated : 0;
}

/**
 * Compute structure match between original control image and generated image
 *
 * @param originalControlBuffer - The original high-contrast control image (white on black)
 * @param generatedImageUrl - URL of the AI-generated image
 * @returns StructureMatchResult with scores and classification
 */
export async function computeStructureMatch(
  originalControlBuffer: Buffer,
  generatedImageUrl: string
): Promise<StructureMatchResult> {
  try {
    logger.debug('[StructureMatch] Downloading generated image...');
    const generatedBuffer = await downloadImage(generatedImageUrl);

    logger.debug('[StructureMatch] Extracting masks...');
    const targetSize = 512;  // Smaller for faster comparison

    // Extract original mask (simple threshold - it's already high contrast)
    const originalResult = await extractBinaryMask(originalControlBuffer, 128, targetSize);

    // Extract generated mask (adaptive threshold for textured image)
    const generatedResult = await extractGeneratedSigilMask(generatedBuffer, targetSize);

    logger.debug('[StructureMatch] Computing IoU...');
    const iouResult = computeIoU(originalResult.mask, generatedResult.mask);

    logger.debug('[StructureMatch] Computing edge overlap...');
    const edgeOverlap = computeEdgeOverlap(
      originalResult.mask,
      generatedResult.mask,
      originalResult.width,
      originalResult.height,
      3  // 3px tolerance
    );

    // Weighted combination (IoU is stricter, edge overlap is more forgiving)
    const combinedScore = 0.7 * iouResult.iou + 0.3 * edgeOverlap;

    // Classify
    let classification: 'Structure Preserved' | 'More Artistic' | 'Style Drift';
    if (combinedScore >= THRESHOLDS.preserved) {
      classification = 'Structure Preserved';
    } else if (combinedScore >= THRESHOLDS.artistic) {
      classification = 'More Artistic';
    } else {
      classification = 'Style Drift';
    }

    const result: StructureMatchResult = {
      iouScore: iouResult.iou,
      edgeOverlapScore: edgeOverlap,
      combinedScore,
      structurePreserved: combinedScore >= THRESHOLDS.preserved,
      classification,
      analysisDetails: {
        originalWhitePixels: iouResult.mask1Pixels,
        generatedWhitePixels: iouResult.mask2Pixels,
        intersectionPixels: iouResult.intersection,
        unionPixels: iouResult.union,
      },
    };

    logger.info('[StructureMatch] Complete', {
      iouScore: result.iouScore.toFixed(3),
      edgeOverlap: result.edgeOverlapScore.toFixed(3),
      combined: result.combinedScore.toFixed(3),
      classification: result.classification,
    });

    return result;

  } catch (error) {
    logger.error('[StructureMatch] Error computing structure match', error);

    // Return conservative estimate on error (assume drift)
    return {
      iouScore: 0.5,
      edgeOverlapScore: 0.5,
      combinedScore: 0.5,
      structurePreserved: false,
      classification: 'More Artistic',
      analysisDetails: {
        originalWhitePixels: 0,
        generatedWhitePixels: 0,
        intersectionPixels: 0,
        unionPixels: 0,
      },
    };
  }
}

/**
 * Batch compute structure match for multiple generated images
 */
export async function computeStructureMatchBatch(
  originalControlBuffer: Buffer,
  generatedImageUrls: string[]
): Promise<StructureMatchResult[]> {
  const results = await Promise.all(
    generatedImageUrls.map(url => computeStructureMatch(originalControlBuffer, url))
  );
  return results;
}

File: .git-rewrite/t/ai-service/src/structure_matching.py
"""
Anchor AI Service - Structure Matching Module
Computes structure preservation metrics between original sigil and generated output.

Key metrics:
- IoU (Intersection over Union) of binarized masks
- Edge overlap score using Canny edge detection
- Combined structure match score
"""

import numpy as np
from PIL import Image
import cv2
from typing import Tuple, Optional
from dataclasses import dataclass

from .config import structure_match_config, StructureMatchConfig


@dataclass
class StructureMatchResult:
    """Result from structure matching analysis."""

    iou_score: float                    # Intersection over Union (0-1)
    edge_overlap_score: float           # Edge-based overlap (0-1)
    combined_score: float               # Weighted combination (0-1)
    structure_preserved: bool           # True if above threshold
    classification: str                 # "Structure Preserved" or "Style Drift" or "More Artistic"
    analysis: dict                      # Detailed analysis info


def binarize_image(
    image: Image.Image,
    threshold: int = 128,
    invert: bool = False
) -> np.ndarray:
    """
    Convert image to binary mask.

    Args:
        image: Input image (grayscale or RGB)
        threshold: Binarization threshold (0-255)
        invert: If True, invert the result

    Returns:
        Binary numpy array (0 or 255)
    """
    # Convert to grayscale if needed
    if image.mode != 'L':
        gray = image.convert('L')
    else:
        gray = image

    # Convert to numpy
    img_array = np.array(gray)

    # Binarize
    _, binary = cv2.threshold(img_array, threshold, 255, cv2.THRESH_BINARY)

    if invert:
        binary = 255 - binary

    return binary


def extract_sigil_from_generated(
    generated_image: Image.Image,
    method: str = "adaptive"
) -> np.ndarray:
    """
    Extract sigil structure from AI-generated image.

    Generated images may have complex textures/colors. This function
    attempts to isolate the sigil structure.

    Args:
        generated_image: AI-generated image
        method: Extraction method ("adaptive", "otsu", "edges")

    Returns:
        Binary mask of detected sigil structure
    """
    # Convert to grayscale
    gray = np.array(generated_image.convert('L'))

    if method == "adaptive":
        # Adaptive thresholding works well for varied backgrounds
        binary = cv2.adaptiveThreshold(
            gray, 255, cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
            cv2.THRESH_BINARY, 21, 5
        )
        # May need to invert depending on sigil color
        if np.mean(binary) > 127:
            binary = 255 - binary

    elif method == "otsu":
        # Otsu's method finds optimal threshold
        _, binary = cv2.threshold(gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)
        if np.mean(binary) > 127:
            binary = 255 - binary

    elif method == "edges":
        # Use edge detection
        edges = cv2.Canny(gray, 50, 150)
        # Dilate to connect edges
        kernel = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (3, 3))
        binary = cv2.dilate(edges, kernel, iterations=1)

    else:
        # Simple threshold
        _, binary = cv2.threshold(gray, 128, 255, cv2.THRESH_BINARY)

    return binary


def compute_iou(
    mask1: np.ndarray,
    mask2: np.ndarray
) -> Tuple[float, dict]:
    """
    Compute Intersection over Union between two binary masks.

    IoU = (Intersection Area) / (Union Area)

    Args:
        mask1: First binary mask
        mask2: Second binary mask

    Returns:
        Tuple of (IoU score 0-1, analysis dict)
    """
    # Ensure binary
    mask1_binary = (mask1 > 127).astype(np.uint8)
    mask2_binary = (mask2 > 127).astype(np.uint8)

    # Compute intersection and union
    intersection = np.logical_and(mask1_binary, mask2_binary).sum()
    union = np.logical_or(mask1_binary, mask2_binary).sum()

    if union == 0:
        # Both masks are empty
        return 1.0, {"intersection": 0, "union": 0, "empty_masks": True}

    iou = intersection / union

    analysis = {
        "intersection_pixels": int(intersection),
        "union_pixels": int(union),
        "mask1_pixels": int(mask1_binary.sum()),
        "mask2_pixels": int(mask2_binary.sum()),
    }

    return float(iou), analysis


def compute_edge_overlap(
    image1: np.ndarray,
    image2: np.ndarray,
    canny_low: int = 50,
    canny_high: int = 150,
    tolerance_px: int = 3
) -> Tuple[float, dict]:
    """
    Compute edge-based overlap between two images.

    Uses Canny edge detection and allows for small positional tolerance.

    Args:
        image1: First image (grayscale)
        image2: Second image (grayscale)
        canny_low: Canny low threshold
        canny_high: Canny high threshold
        tolerance_px: Pixel tolerance for edge matching

    Returns:
        Tuple of (overlap score 0-1, analysis dict)
    """
    # Ensure grayscale
    if len(image1.shape) == 3:
        image1 = cv2.cvtColor(image1, cv2.COLOR_RGB2GRAY)
    if len(image2.shape) == 3:
        image2 = cv2.cvtColor(image2, cv2.COLOR_RGB2GRAY)

    # Detect edges
    edges1 = cv2.Canny(image1, canny_low, canny_high)
    edges2 = cv2.Canny(image2, canny_low, canny_high)

    # Dilate edges slightly for tolerance
    if tolerance_px > 0:
        kernel = cv2.getStructuringElement(
            cv2.MORPH_ELLIPSE,
            (tolerance_px * 2 + 1, tolerance_px * 2 + 1)
        )
        edges1_dilated = cv2.dilate(edges1, kernel, iterations=1)
        edges2_dilated = cv2.dilate(edges2, kernel, iterations=1)
    else:
        edges1_dilated = edges1
        edges2_dilated = edges2

    # Compute overlap: edges from image1 that fall within dilated edges2
    edge1_pixels = (edges1 > 0).sum()
    edge2_pixels = (edges2 > 0).sum()

    if edge1_pixels == 0 or edge2_pixels == 0:
        return 0.0, {"edge1_pixels": int(edge1_pixels), "edge2_pixels": int(edge2_pixels), "no_edges": True}

    # How many edges from original are covered by generated
    forward_match = np.logical_and(edges1 > 0, edges2_dilated > 0).sum()
    forward_ratio = forward_match / edge1_pixels

    # How many edges from generated are covered by original
    backward_match = np.logical_and(edges2 > 0, edges1_dilated > 0).sum()
    backward_ratio = backward_match / edge2_pixels

    # Use harmonic mean (F1-style) for balanced metric
    if forward_ratio + backward_ratio > 0:
        edge_score = 2 * forward_ratio * backward_ratio / (forward_ratio + backward_ratio)
    else:
        edge_score = 0.0

    analysis = {
        "edge1_pixels": int(edge1_pixels),
        "edge2_pixels": int(edge2_pixels),
        "forward_match": int(forward_match),
        "forward_ratio": float(forward_ratio),
        "backward_match": int(backward_match),
        "backward_ratio": float(backward_ratio),
    }

    return float(edge_score), analysis


def resize_to_match(
    image1: np.ndarray,
    image2: np.ndarray
) -> Tuple[np.ndarray, np.ndarray]:
    """Resize images to match dimensions."""
    h1, w1 = image1.shape[:2]
    h2, w2 = image2.shape[:2]

    if h1 != h2 or w1 != w2:
        # Resize to the smaller dimensions
        target_h = min(h1, h2)
        target_w = min(w1, w2)
        image1 = cv2.resize(image1, (target_w, target_h))
        image2 = cv2.resize(image2, (target_w, target_h))

    return image1, image2


def compute_structure_match(
    original_mask: Image.Image | np.ndarray,
    generated_image: Image.Image | np.ndarray,
    config: Optional[StructureMatchConfig] = None,
    extraction_method: str = "adaptive"
) -> StructureMatchResult:
    """
    Main function to compute structure preservation score.

    Compares original sigil mask with AI-generated output to determine
    if the structure was preserved.

    Args:
        original_mask: Binary mask of original sigil (white strokes on black)
        generated_image: AI-generated output image
        config: Structure matching configuration
        extraction_method: Method to extract sigil from generated image

    Returns:
        StructureMatchResult with all scores and analysis
    """
    if config is None:
        config = structure_match_config

    # Convert to numpy arrays
    if isinstance(original_mask, Image.Image):
        original_array = np.array(original_mask.convert('L'))
    else:
        original_array = original_mask
        if len(original_array.shape) == 3:
            original_array = cv2.cvtColor(original_array, cv2.COLOR_RGB2GRAY)

    if isinstance(generated_image, Image.Image):
        generated_array = np.array(generated_image.convert('L'))
        generated_rgb = np.array(generated_image.convert('RGB'))
    else:
        if len(generated_image.shape) == 3:
            generated_rgb = generated_image
            generated_array = cv2.cvtColor(generated_image, cv2.COLOR_RGB2GRAY)
        else:
            generated_array = generated_image
            generated_rgb = cv2.cvtColor(generated_image, cv2.COLOR_GRAY2RGB)

    # Resize to match if needed
    original_array, generated_array = resize_to_match(original_array, generated_array)
    _, generated_rgb = resize_to_match(original_array, generated_rgb)

    # Binarize original mask
    original_binary = binarize_image(
        Image.fromarray(original_array),
        threshold=config.binarize_threshold
    )

    # Extract sigil structure from generated
    generated_binary = extract_sigil_from_generated(
        Image.fromarray(generated_rgb),
        method=extraction_method
    )

    # Compute IoU
    iou_score, iou_analysis = compute_iou(original_binary, generated_binary)

    # Compute edge overlap
    edge_score, edge_analysis = compute_edge_overlap(
        original_array,
        generated_array,
        tolerance_px=3
    )

    # Compute combined score
    combined_score = (
        config.iou_weight * iou_score +
        config.edge_weight * edge_score
    )

    # Determine classification
    structure_preserved = combined_score >= config.iou_threshold

    if combined_score >= 0.90:
        classification = "Structure Preserved"
    elif combined_score >= config.iou_threshold:
        classification = "Structure Preserved"
    elif combined_score >= 0.70:
        classification = "More Artistic"
    else:
        classification = "Style Drift"

    # Compile analysis
    analysis = {
        "iou_analysis": iou_analysis,
        "edge_analysis": edge_analysis,
        "thresholds": {
            "iou_threshold": config.iou_threshold,
            "edge_threshold": config.edge_match_threshold,
        },
        "extraction_method": extraction_method,
    }

    return StructureMatchResult(
        iou_score=iou_score,
        edge_overlap_score=edge_score,
        combined_score=combined_score,
        structure_preserved=structure_preserved,
        classification=classification,
        analysis=analysis
    )


def compute_batch_structure_match(
    original_mask: Image.Image | np.ndarray,
    generated_images: list[Image.Image | np.ndarray],
    config: Optional[StructureMatchConfig] = None
) -> list[StructureMatchResult]:
    """
    Compute structure match for multiple generated variations.

    Args:
        original_mask: Binary mask of original sigil
        generated_images: List of AI-generated images
        config: Structure matching configuration

    Returns:
        List of StructureMatchResult for each image
    """
    return [
        compute_structure_match(original_mask, img, config)
        for img in generated_images
    ]


def should_regenerate(
    results: list[StructureMatchResult],
    min_passing: int = 2
) -> Tuple[bool, list[int]]:
    """
    Determine if regeneration is needed based on structure match results.

    Args:
        results: Structure match results for all variations
        min_passing: Minimum number of variations that should pass

    Returns:
        Tuple of (should_regenerate, list of passing indices)
    """
    passing_indices = [
        i for i, result in enumerate(results)
        if result.structure_preserved
    ]

    should_regen = len(passing_indices) < min_passing

    return should_regen, passing_indices
